packages/hardhat-core/src/common/bigInt.ts

```diff
@@ -0,0 +1,97 @@
+import type { BigNumber as EthersBigNumberType } from "ethers";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import type { BigNumber as BigNumberJsType } from "bignumber.js";
+// eslint-disable-next-line import/no-extraneous-dependencies
+import type { default as BNType } from "bn.js";
+
+export function normalizeToBigInt(
+  source:
+    | number
+    | bigint
+    | BNType
+    | EthersBigNumberType
+    | BigNumberJsType
+    | string
+): bigint {
+  if (isBigNumber(source)) {
+    return BigInt(source.toString());
+  } else if (
+    typeof source === "string" ||
+    typeof source === "number" ||
+    typeof source === "bigint"
+  ) {
+    if (typeof source === "number") {
+      if (!Number.isInteger(source)) {
+        // eslint-disable-next-line @nomiclabs/hardhat-internal-rules/only-hardhat-error
+        throw new RangeError(
```

Comment:
We have to use a `HardhatError` here.

Reply:
Addressed in https://github.com/NomicFoundation/hardhat/pull/2428/commits/af03a9a678c475c62e4e6f0a93db87c2ce901ba5